lib/oper1.g

```diff
@@ -360,6 +360,40 @@ BIND_GLOBAL( "InstallOtherMethod",
     end );
 
 
+#############################################################################
+##
+#F  InstallEarlyMethod( <opr>,<method> )
+##
+##  <#GAPDoc Label="InstallEarlyMethod">
+##  <ManSection>
+##  <Func Name="InstallEarlyMethod" Arg="opr,method"/>
+##
+##  <Description>
+##  installs a special "early" function method <A>method</A> for the
+##  operation <A>opr</A>. An early method is special in that it bypasses
+##  method dispatch, and is always the first method to be called when
+##  invoking the operation. This can be used to avoid method selection
+##  overhead for certain special cases.
+##  <P/>
+##  Unlike with regular methods, no checks are performed on the
+##  arguments. Hence any operation can have at most one such early
+##  method for each arity (i.e., one early method taking 1 argument, one
+##  early method taking 2 arguments, etc.).
+##  <P/>
+##  If an early method detects that it is not applicable, it can resume
+##  regular method dispatch by invoking <Ref Func="TryNextMethod"/>.
+##  <P/>
+##  For example, we install early methods for <Ref Oper="First"/> that
+##  deal with internal lists, for which computing their types (and hence
+##  method selection) can be very expensive.
+##  </Description>
+##  </ManSection>
+##  <#/GAPDoc>
+##
+BIND_GLOBAL( "InstallEarlyMethod", INSTALL_EARLY_METHOD );
+#TODO; store location info somehow
```

Comment:
Perhaps `InstallEarlyMethod` should also take an (optional) `info` string, so that one can e.g. write something like `for plain lists and a function`.

But then that also has to be stored somewhere. One solution for that: change `OPER(oper)->earlyMethod[n]` to point to a plist which contains the function, the location of the installation, and the info string. The result will have identical overhead to what is now in this PR when there is *no* early method; and only a small extra overhead in the rare case where there is one.

Thoughts on this, anyone?

Reply:
We could also add components `earlyMethodInfoString` and `earlyMethodInstallLocation`.

Reply:
I would be happy to just call these "early method", rather than record an extra string.

Reply:
they currently *are* called "early method", I just thought it might be nicer to see something like "early method for a plain list and a function"; given that no predicates are shown for an early method, so without this, the only way to have any idea what it does is to inspect its code. But yeah, it's certainly a secondary feature, just like tracking the location of the `InstallEarlyMethod` call (useful for debugging, but not crucial for the feature overall).